In [ ]:
!apt-get moo

                 (__) 
                 (oo) 
           /------\/ 
          / |    ||   
         *  /\---/\ 
            ~~   ~~   
..."Have you mooed today?"...


В этом ноутбуке мы рассмотрим базовые команды командной строки в Linux. Данный ноутбук не является исчерпывающим описание того, как работать с операционной системой Unix-based, но позволяет сориентироваться в некоторых деталях устройства операционной системы и команд.


# Команды в UNIX

Линукс - самый известный пример операционной системы из большого семейства операционных систем под названием UNIX, которые происходят от ОС AT&T Unix, разработанной еще в 1970-х.

В операционных системах UNIX-типа для взаимодействия с ОС используют специальные команды. До появления привычного оконного интерфейса пользователи взаимодействовали с компьютером исключительно с помощью этих команд.

Сейчас этот способ используют и сейчас, т. к. у него есть ряд преимуществ:
 - он удобнее при подключение при управление ОС через сеть;
 - из команды можно составлять скрипты для автоматизации задач.

 Если вам необходимо решить сложную вычислительную задачу с использованием удалённых машин (например, вычислить что-то на кластере), вам так же придётся использовать команды UNIX-систем для управления вашими вычислениями.



Ключевые понятия:
- **Командная строка** - интерфейс пользователя, окно для вывода системных сообщений и приёма команд;
- **Терминал** - программа, которая отображает строку. Например: *gnome-terminal*, *xterm*, *tilda*;
- **Оболочка (shell)** - программа которая выполняет команды. Например: *bash*, *zsh*, *sh*;
- **Консоль** - комплект устройств интерактивного ввода-вывода (дисплей, клавиатура, мышь) для управления компьютером.

Запускать команды можно как из терминала, так и из юпитера или колаба (Google Colab), поставив перед командой "!":

In [ ]:
!echo "Hello World"

Hello World


# Файловая система

В Unix-системах (и не только в них) все файлы образуют структуру похожую на дерево, в каждой директории (папке) хранятся файлы и другие директории. Самая первая директория которая содержит все остальные, называется корневой (*root*)


У каждой директории есть путь (по сути, просто строка) который показывает где она находится в этом дереве. "/" - это путь корневой директории.

---



![alt text](http://linuxcommand.org/images/file_manager.jpg)

В командной строке все команды выполняются в какой-либо директории.

Представьте, что дерево файловой системы - это здание, и каждая директория - это комната с файлами и дверьми в другие директории. В каждый момент времени вы находитесь в одной комнате.

Как можно перемещаться по зданию?

Обратие внимание, все команды образованы от англоязычных словосочетаний и потому могут легко запомниаться при наличии простых знаний английского языка.

In [ ]:
# Вывести местоположение, путь до текущей директории
!pwd

/content


In [ ]:
# Посмотрим, что в нашей директории есть
!ls
#ls is for "list"

sample_data


In [ ]:
#Создадим пустой файлик
!touch test-file
# пробел в имени - плохая идея, создадутся два файла
!touch hello world

# посмотрим что получилось
!ls

hello  sample_data  test-file  world


In [ ]:
#Удаление файлов
!rm world
!ls

hello  sample_data  test-file


In [ ]:
# Новая директория
!mkdir please_come_here


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если знаешь, что делаешь.

Это так называемые скрытые директории (hidden или dot directories/files) - их имя начинается с точки.


In [ ]:
# Покажем больше информации и отобразим скрытые файлы
!ls -al
# Первое слово - команда, а второе - аргумент

total 20
drwxr-xr-x 1 root root 4096 Nov  6 15:18 .
drwxr-xr-x 1 root root 4096 Nov  6 15:10 ..
drwxr-xr-x 4 root root 4096 Nov  3 18:00 .config
-rw-r--r-- 1 root root    0 Nov  6 15:18 hello
drwxr-xr-x 2 root root 4096 Nov  6 15:18 please_come_here
drwxr-xr-x 1 root root 4096 Nov  3 18:00 sample_data
-rw-r--r-- 1 root root    0 Nov  6 15:18 test-file




```
.  - это путь к той директории в которой мы уже находимся
.. - это путь к родительской директории
.config - это скрытая папка
```

Каждая строчка - это одна из "дверей" в другую директорию либо файл.
*Относительный* путь это не полный "адрес" директории а то, куда нам нужно пойти из места в котором мы сейчас находимся.


```
 drwxr-xr-x 1 root root 4096 Feb 15 17:21 please_come_here
 |                           дата создания    |
 |                                         _относительный_ путь
 |
  \ d - эта метка значит, что это путь к другой директории (d for directory)
```



In [ ]:
# переход в другую директорию команда cd
!pwd
!cd please_come_here
!pwd

/content
/content


Заработало?

Не должно. Тут проблема в том что юпитер при исполнении каждой команды запускает новую оболочку в своей директории (той, откуда его вызвали), и после завершения команды она закрывается. Т. е. в одной оболочке он выполнил команду pwd, закрыл её, в другой выполнил команду cd please_come_here, закрыл её, и в третьей выполнил команду pwd и закрыл её.

Чтоб команды запускались в новой директории нужно поменять текущую директорию самого юпитера с помощью команды:

```
%cd
```


In [ ]:
!pwd
%cd please_come_here
!pwd

/content
/content/please_come_here
/content/please_come_here


Если выполнить команду cd без аргументов - он вернёт вас в домашнюю директорию:

In [ ]:
%cd
!pwd

/root
/root


Когда мы переходим в другую директорию, ее имя добавляется к нашему пути через "/".

Например, были в `/home`, перешли в директорию `Pictures `оказались в `/home/Pictures`.

Абсолютный путь файла в этой директории будет `/home/Picutes/image.png`.

Если мы хотим взять файл из директории, которая является дочерней к той директории, где мы сейчас находимся, мы можем указать относительный путь файла.

Например, из директории `/home` мы можем обратиться к файлу по пути `Pictures/image.png`.

In [ ]:
# Можно выполнить несколько команд последовательно:
# с помощью символа ; - для безусловного последовательного выполнения команд, они выполнятся последовательно одна за другой:
!pwd; cd please_come_here; pwd

# с помощью символа && - при этом вторая (и последующая) команда выполнится только в случае успешного выполнения первой (предыдущей):
!pwd && cd please_come_here && pwd

# с помощью символа || - при этом вторая (и последующая) команда выполнится только в случае неуспешного выполнения первой (предыдущей):
!pwd || cd please_come_here && pwd


/root
/bin/bash: line 1: cd: please_come_here: No such file or directory
/root
/root
/bin/bash: line 1: cd: please_come_here: No such file or directory
/root
/root


## Скачивание из интернета

In [ ]:
# сохраняем страницу по адресу

!curl example.com >> example-com.html
!ls
!cat example-com.html


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1256  100  1256    0     0  46229      0 --:--:-- --:--:-- --:--:-- 46518
English.txt  example-com.html
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-radius: 0.5em;
        box-shadow: 2px 3px 7px 2px rgba(0,0,0,0.02);
    }
    a:link, a:vis

In [ ]:
!curl instagram.com

Иногда сервера пытаются перенаправить запрос на другой адрес. Curl не следует по таким перенаправлениям. Есть програмка wget которая может переходить по таким ссылкам и вообще проходить по всем ссылкам на странице и скачивать их.

In [ ]:
!wget instagram.com -O insta.html

--2023-11-06 15:20:32--  http://instagram.com/
Resolving instagram.com (instagram.com)... 157.240.254.174, 2a03:2880:f275:1e9:face:b00c:0:4420
Connecting to instagram.com (instagram.com)|157.240.254.174|:80... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://instagram.com/ [following]
--2023-11-06 15:20:32--  https://instagram.com/
Connecting to instagram.com (instagram.com)|157.240.254.174|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://www.instagram.com/ [following]
--2023-11-06 15:20:32--  https://www.instagram.com/
Resolving www.instagram.com (www.instagram.com)... 157.240.254.174, 2a03:2880:f275:1e9:face:b00c:0:4420
Connecting to www.instagram.com (www.instagram.com)|157.240.254.174|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.facebook.com/unsupportedbrowser [following]
--2023-11-06 15:20:32--  https://www.facebook.com/unsupportedbrowser
Resolv

Обратите внимание на

`HTTP request sent, awaiting response... 301 Moved Permanently` (четвёртая строка)


# Аргументы и параметры к командам

Обычно **аргументы** – это то, над чем выполняется команда. Например, команда перехода в другую директорию должна получить в качестве аргумента путь директории, в которую следует перейти. Команде вывода на экран содержимого файла надо передать имя файла.

**Параметры** модифицируют, как бы видоизменяют/управляют работой программы. Например, с помощью команды просмотра содержимого каталога можно вывести простой список вложенных объектов. Используя же определенный параметр, можно получить подробные сведения о каждом файле и папке.

Аргументы указывают через пробел, как отдельные слова.

Параметры указывают после дефиса: `-f` (однобуквенные) либо после двух дефисов: `--file`.

In [ ]:
# Рассмотрим код на примере утилиты "календарь". Для начала её скачаем и установим.

!sudo apt-get -y install ncal

Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following NEW packages will be installed:
  ncal
0 upgraded, 1 newly installed, 0 to remove and 19 not upgraded.
Need to get 20.2 kB of archives.
After this operation, 69.6 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/universe amd64 ncal amd64 12.1.7+nmu3ubuntu2 [20.2 kB]
Fetched 20.2 kB in 0s (90.2 kB/s)
debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 78, <> line 1.)
debconf: falling back to frontend: Readline
debconf: unable to initialize frontend: Readline
debconf: (This frontend requires a controlling tty.)
debconf: falling back to frontend: Teletype
dpkg-preconfigure: unable to re-open stdin: 
Selecting previously unselected package ncal.
(Reading database ... 120874 files and directories curr

In [ ]:
# календарик на весь год

!ncal 2023

                                  2023
    January           February          March             April             
Su  1  8 15 22 29        5 12 19 26        5 12 19 26        2  9 16 23 30
Mo  2  9 16 23 30        6 13 20 27        6 13 20 27        3 10 17 24   
Tu  3 10 17 24 31        7 14 21 28        7 14 21 28        4 11 18 25   
We  4 11 18 25        1  8 15 22        1  8 15 22 29        5 12 19 26   
Th  5 12 19 26        2  9 16 23        2  9 16 23 30        6 13 20 27   
Fr  6 13 20 27        3 10 17 24        3 10 17 24 31        7 14 21 28   
Sa  7 14 21 28        4 11 18 25        4 11 18 25        1  8 15 22 29   

    May               June              July              August            
Su     7 14 21 28        4 11 18 25        2  9 16 23 30     6 13 20 27   
Mo  1  8 15 22 29        5 12 19 26        3 10 17 24 31     7 14 21 28   
Tu  2  9 16 23 30        6 13 20 27        4 11 18 25     1  8 15 22 29   
We  3 10 17 24 31        7 14 21 28        5 12 19 26   

In [ ]:
# передаем 2 аргумента

!ncal 11 2023


    November 2023     
Su     5 12 19 26   
Mo     6 13 20 27   
Tu     7 14 21 28   
We  1  8 15 22 29   
Th  2  9 16 23 30   
Fr  3 10 17 24      
Sa  4 11 18 25      


In [ ]:
# передаем параметр -b

!ncal -b 11 2023


   November 2023      
Su Mo Tu We Th Fr Sa  
          1  2  3  4  
 5  6  7  8  9 10 11  
12 13 14 15 16 17 18  
19 20 21 22 23 24 25  
26 27 28 29 30        
                      


In [ ]:
# Однобуквенные параметры можно склеивать

!ncal -b -w 11 2023
!ncal -bw 11 2023

      November 2023        
 w| Su Mo Tu We Th Fr Sa   
44|           1  2  3  4   
45|  5  6  7  8  9 10 11   
46| 12 13 14 15 16 17 18   
47| 19 20 21 22 23 24 25   
48| 26 27 28 29 30         
                           
      November 2023        
 w| Su Mo Tu We Th Fr Sa   
44|           1  2  3  4   
45|  5  6  7  8  9 10 11   
46| 12 13 14 15 16 17 18   
47| 19 20 21 22 23 24 25   
48| 26 27 28 29 30         
                           
